### PR TITLE
Ees 1830 add new observation tables with identity columns

### PR DESCRIPTION
**Pull request by duncan-at-hiveit**
This PR:

- adds in 2 new tables, ObservationRow and ObservationRowFilterItem which mirror Observation and ObservationFilterItem but with Identity ids
- adds a new stored procedure that migrates batches of Subjects from the existing tables to the new
- sets up a new Data Factory trigger that will migrate batches of Subjects from Observation and ObservationFilterItem into ObservationRow and ObservationRowFilterItem via the stored procedure

### CODE REVIEW

**Review by benoutram — APPROVED** [review]

**Comment by benoutram** [comment]
Adding Do Not Merge label while the Infrastructure build is failing due to another issue.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20210125164907_CreateNewObservationTables.sql [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+-- 
+-- create index IX_ObservationRow_TimeIdentifier
+--     on ObservationRow (TimeIdentifier)
+-- go
+-- 
+-- create index IX_ObservationRow_Year
+--     on ObservationRow (Year)
+-- go
+-- 
+-- create index NCI_WI_ObservationRow_SubjectId
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20210125164907_CreateNewObservationTables.sql:48]: That name was autogenerated by the Azure performance tuning tool in the portal. `NCI` is short for non-clustered index. Not sure what the `WI` is.  I've left a comment about this on https://dfedigital.atlassian.net/browse/EES-1836 @duncan-at-hiveit .

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20201120140950_Routine_RemoveSoftDeletedSubjects.sql [hunk 2]
```diff
@@ -68,13 +68,20 @@ BEGIN
                                      WHERE Filter.Id = OFI.FilterId
                                        AND Filter.SubjectId IN (SELECT Id FROM #SoftDeletedSubjects));
 
+                        DELETE TOP (@ObservationFilterItemCommitBatchSize) OFI
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20201120140950_Routine_RemoveSoftDeletedSubjects.sql:71]: You are missing a migration for this change.

**Review by benoutram — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20201120140950_Routine_RemoveSoftDeletedSubjects.sql (hunk 2 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20201120140950_Routine_RemoveSoftDeletedSubjects.sql:71]: Hey @benoutram, not sure what you think might be missing?  The ObservationRowFilterItem table is getting created in the "20210125164907_EES1830_AddNewObservationTables.cs" migration, unless you were referring to something else?

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20210125164907_CreateNewObservationTables.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20210125164907_CreateNewObservationTables.sql:48]: Yeah @ntsim I was just re-including the existing indexes that existed on the Observation and ObservationRow table here - it's generated from SSMS "Create Table" scripts.  Happy to change on 1836 if we'd prefer a different convention though - any thoughs @benoutram @ntsim ?  Totally happy to go with `IX_ObservationRow_SubjectId`!

[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20210125164907_CreateNewObservationTables.sql:48]: > That name was autogenerated by the Azure performance tuning tool in the portal. `NCI` is short for non-clustered index. Not sure what the `WI` is. I've left a comment about this on https://dfedigital.atlassian.net/browse/EES-1836 @duncan-at-hiveit .

No worries thanks @benoutram - I've created a ticket https://dfedigital.atlassian.net/browse/EES-1856 for this subtask, thanks!

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20201120140950_Routine_RemoveSoftDeletedSubjects.sql (hunk 2 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20201120140950_Routine_RemoveSoftDeletedSubjects.sql:71]: I can't see what is reloading `20201120140950_Routine_RemoveSoftDeletedSubjects.sql` i.e. updadting  `RemoveSoftDeletedSubjects`, and as you know at the moment the convention is to duplicate the source and alter it, although I agree with your suggestion in future we could have a _current_ and _previous_ for all of them.

**Review by benoutram — APPROVED** [review]

**Pull request merged**